storage.go

```diff
@@ -325,6 +352,68 @@ func (s *Storage) keyList() (map[string]bool, uint64, error) {
 	}
 	return keyList, rev, nil
 }
+
+// keyListAdd adds the given key to the keylist
+func (s *Storage) keyListAdd(key string) error {
+	keyList, keyListRev, err := s.keyList()
+	if err != nil && !errNotFound(err) {
+		return err
+	}
+	if keyList == nil {
+		keyList = make(map[string]bool)
+	}
+	// If the key is already in the keylist there's nothing to do.
+	if _, exists := keyList[key]; exists {
+		return nil
+	}
+	keyList[key] = true
+	bytes, err := json.Marshal(keyList)
+	if err != nil {
+		return errors.AddContext(err, "failed to serialise the new key list")
+	}
+	err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
+	if err != nil {
+		return errors.AddContext(err, "failed to store the key list")
+	}
+	return nil
+}
+
+// keyListDelete deletes the given key from the keylist
+func (s *Storage) keyListDelete(key string) error {
+	keyList, keyListRev, err := s.keyList()
+	if err != nil && !errNotFound(err) {
+		return err
+	}
+	// If the keylist is empty there's nothing to do.
+	if keyList == nil {
+		return nil
+	}
+	// If the key is not in the keylist there's nothing to do.
+	if _, exists := keyList[key]; !exists {
+		return nil
+	}
+	delete(keyList, key)
+	bytes, err := json.Marshal(keyList)
+	if err != nil {
+		return errors.AddContext(err, "failed to serialise the new key list")
+	}
+	err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
+	if err != nil {
+		return errors.AddContext(err, "failed to store the key list")
+	}
+	return nil
+}
+
+// errNotFound checks the various failure modes of the registry which all mean
+// that the entry was not found.
+func errNotFound(err error) bool {
+	// This check covers all cases:
+	// - skydb.ErrNotFound
+	// - renter.ErrRegistryEntryNotFound
+	// - renter.ErrRegistryLookupTimeout
+	return err != nil && (strings.Contains(err.Error(), "entry not found"))
```

Comment:
Could make this `return err != nil && (strings.Contains(err.Error(), ErrNotFound.Err()))`

Reply:
Not really. There are two reasons:  
1. `ErrNotFound`'s value is `skydb entry not found` and that's intended - I want to know at which stage in the chain we fail to retrieve the value.  
2. Using any specific error value, rather than a string, would give the reader the wrong impression that that value is somehow the base error we're looking for. That's not the case - `entry not found` just happens to be the common substring among all error messages we want to catch. If that was not the case we'd have something like
```go
return err != nil && (errors.Contains(err, skydb.ErrNotFound) || errors.Contains(err, renter.ErrRegistryEntryNotFound) || errors.Contains(err, renter.ErrRegistryLookupTimeout))
```
We might need to switch to this approach if it turns out that we need to cover another error as well but for the moment this line is very simple, very readable, and it does the job.

Reply:
We ended up doing the check Chris suggested explicitly for every error we want to check for. This is a little verbose but it's quite readable and it's not as fragile as the "shared error text substring" approach above.